contracts/strategy/TradeFintechStrategy.sol

```diff
@@ -0,0 +1,134 @@
+        require(block.timestamp <= depositPeriodEnds, "Deposit period has ended");
+        asset.transfer(managerAddress, _amount);
+        totalInvestedInRWA += _amount;
+    }
+
+    function _freeFunds(uint256 _amount) internal override {
+        require(block.timestamp > lockPeriodEnds, "Lock period has not ended");
+        // We don't check available liquidity because we need the tx to
+        // revert if there is not enough liquidity so we don't improperly
+        // pass a loss on to the user withdrawing.
+        asset.safeTransferFrom(managerAddress, address(this), _amount);
+        totalInvestedInRWA -= _amount;
+    }
+
+    /**
+     * @dev Optional function for a strategist to override that will
+     * allow management to manually withdraw deployed funds from the
+     * yield source if a strategy is shutdown.
+     *
+     * This should attempt to free `_amount`, noting that `_amount` may
+     * be more than is currently deployed.
+     *
+     * NOTE: This will not realize any profits or losses. A separate
+     * {report} will be needed in order to record any profit/loss. If
+     * a report may need to be called after a shutdown it is important
+     * to check if the strategy is shutdown during {_harvestAndReport}
+     * so that it does not simply re-deploy all funds that had been freed.
+     *
+     * EX:
+     *   if(freeAsset > 0 && !TokenizedStrategy.isShutdown()) {
+     *       depositFunds...
+     *    }
+     *
+     * @param _amount The amount of asset to attempt to free.
+     */
+    function _emergencyWithdraw(uint256 _amount) internal override {
+        uint256 amountToWithdraw = Math.min(_amount, totalInvestedInRWA);
+        asset.safeTransferFrom(managerAddress, address(this), amountToWithdraw);
+        asset.transfer(TokenizedStrategy.management(), amountToWithdraw);
+        totalInvestedInRWA -= amountToWithdraw;
+    }
+
+
+    /// @notice Allows the manager to report gains or losses.
+    /// @dev Should be called before calling report() to report the amount of the gain or loss.
+    /// @dev The manager can only report gains or losses.
+    /// @param _gain The amount of the gain.
+    /// @param _loss The amount of the loss.
+    function reportGainOrLoss(uint256 _gain, uint256 _loss) external {
+        require(msg.sender == managerAddress, "Only the manager can report gains or losses");
+
+        if (_gain > 0) {
+            require(_loss == 0, "Cannot report both gain and loss");
+            totalGains += _gain;
+            emit GainReported(_gain);
+        } else if (_loss > 0) {
+            require(_loss <= totalInvestedInRWA, "Cannot report loss more than total invested");
+            totalLosses += _loss;
+            totalInvestedInRWA -= _loss;
+            emit LossReported(_loss);
+        }
+    }
+
+    /// @notice Allows the manager to return the funds.
+    /// @param _amount The amount that is being returned.
+    function returnFunds(uint256 _amount) external {
```

Comment:
Why we don't drop the value of totalInvestedInRWA in case of returning funds?

Reply:
Need to write down all flow for this storage variable - cases when it is increased/decreased

Reply:
Same reason as before, the next time harvestAndReport is triggered the totalInvested will be updated correctly.

Reply:
The flow here is:

1. Strategy manager return funds
2. Funds are sitting idle on the Strategy now
3. Next time harvestAndReport is called (by the report bot) the totalInvested will be updated and funds will be transferred back to the strategy manager to invest

Reply:
So thats the flow for gain? Whats for loss? I am curious because strategy manager reports only loss, gain is just on harvestAndReport, and I don't understand how its valid

Reply:
That's not correct. Strategy manager will also report any gains on the reportGainOrLoss. But now I got what you are thinking as my previous reply. Will decrease totalInvested after any fund is returned to the Strategy.